arangod/IResearch/Geo.cpp

```diff
@@ -21,91 +21,63 @@
 /// @author Andrey Abramov
 ////////////////////////////////////////////////////////////////////////////////
 
-#include "Geo.h"
-
+#include "IResearch/Geo.h"
+#include "IResearch/IResearchCommon.h"
 #include "Geo/GeoJson.h"
 #include "Geo/ShapeContainer.h"
 #include "Logger/LogMacros.h"
-#include "IResearch/IResearchCommon.h"
-#include "velocypack/Builder.h"
-#include "velocypack/Slice.h"
 
-namespace arangodb {
-namespace iresearch {
+#include <s2/s2latlng.h>
 
-bool parseShape(VPackSlice slice, geo::ShapeContainer& shape, bool onlyPoint) {
-  Result res;
-  if (slice.isObject()) {
-    if (onlyPoint) {
-      S2LatLng ll;
-      res = geo::geojson::parsePoint(slice, ll);
+#include <velocypack/Builder.h>
+#include <velocypack/Slice.h>
+
+namespace arangodb::iresearch {
 
-      if (res.ok()) {
-        shape.resetCoordinates(ll);
+template<bool Validation>
+bool parseShape(velocypack::Slice vpack, geo::ShapeContainer& region,
+                bool onlyPoint) {
+  Result r;
+  if (vpack.isObject()) {
+    if (onlyPoint) {
+      S2LatLng latLng;
+      r = geo::json::parsePoint<Validation>(vpack, latLng);
+      if (!Validation || r.ok()) {
```

Comment:
Why check "Validation" here? "r" is always OK in case of Validation false as I see. Maybe just ASSERT is enough?

Reply:
Answered in other place

Reply:
For disjunction this is not true. But anyway  let's leave it as is. This is not performance critical.

Reply:
Ah no. for "!Validation" it is compile time indeed

Reply:
In general I think most critical time validation is FindValidationError probably Contains, vpack double validation, so maybe I can don't use it in with r.ok